dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/tools/ParallelDependentTestFinder.java

```diff
@@ -0,0 +1,541 @@
+ *     dtFinders.add(dtFinder.runDTF());
+ *     dtFinder = dtFinder.createFromDTFinder(dtFinder, nextDependentTestName);
+ * }
+ *
+ * When done:
+ *
+ * List<String> allDtList = DependentTestFinder.dependenciesToString(dtFinders);
+ *
+ * // Optional write to file:
+ * DependentTestFinder.writeToFile(dtFinders, FILE);
+ * DependentTestFinder.writeToFile(dtFinders); // Writes to file stored in static field DT_FILE
+ *
+ * </pre>
+ */
+public class ParallelDependentTestFinder {
+    private static File DT_FILE;
+
+    private class TestOrder {
+        private final List<String> testOrder;
+        private final Map<String, RESULT> results;
+
+        public TestOrder(final List<String> testOrder, final Map<String, RESULT> results) {
+            this.testOrder = testOrder;
+            this.results = results;
+        }
+
+        public TestOrder(final List<String> testOrder) {
+            this.testOrder = testOrder;
+
+            results = runTestOrder(testOrder).getNameToResultsMap();
+        }
+
+        public RESULT getResult(final String testName) {
+            return results.get(testName);
+        }
+
+        public List<String> getTestsBefore(final String testName) {
+            return testOrder.subList(0, testOrder.indexOf(testName));
+        }
+    }
+
+    private static TestExecResult runTestOrder(final List<String> order) {
+        return new FixedOrderRunner(order).run().getExecutionRecords().get(0);
+    }
+
+    private final String dependentTestName;
+    private final RESULT dependentTestResult;
+    private final TestOrder originalOrder;
+    private final TestOrder newOrder;
+    private final TestOrder primeOrder;
+
+    private final List<String> filesToDelete;
+
+    private final Map<String, TestData> knownDependencies;
+    private final List<TestData> debugLog;
+
+    public ParallelDependentTestFinder(final String dependentTestName,
+                                       final List<String> originalOrder,
+                                       final List<String> newOrder,
+                                       final List<String> filesToDelete) {
+        this.dependentTestName = dependentTestName;
+
+        this.originalOrder = new TestOrder(originalOrder);
+        this.newOrder = new TestOrder(newOrder);
+
+        primeOrder = new TestOrder(generatePrimeOrder(this.originalOrder, this.newOrder));
+
+        this.filesToDelete = filesToDelete;
+        knownDependencies = new HashMap<>();
+
+        dependentTestResult = this.originalOrder.getResult(dependentTestName);
+        debugLog = new ArrayList<>();
+    }
+
+    public ParallelDependentTestFinder(final String dependentTestName,
+                                       final List<String> originalOrder,
+                                       final List<String> newOrder,
+                                       final List<String> filesToDelete,
+                                       final Map<String, TestData> knownDependencies) {
```

Comment:
Perhaps you are still in the middle of changing this implementation but if not, I think the knownDependencies here should be a Map<String, Set<TestData>>. See below and my comment on your writeToFile method earlier today for my explanation why.

"My suggestion would be that the test data class contains a set of before and a set of after tests but your map of precomputed dependences still maps from a dependent test to a set of test data classes. The reason for this implementation is that we can actually generate a "debug" log from this implementation."